File: doordeck-sdk/src/commonMain/kotlin/com/doordeck/multiplatform/sdk/storage/SecureStorage.kt
package com.doordeck.multiplatform.sdk.storage

import com.doordeck.multiplatform.sdk.ApplicationContext
import com.doordeck.multiplatform.sdk.model.data.ApiEnvironment
import kotlin.js.JsExport

/**
 * Interface for secure storage operations.
 */
@JsExport
interface SecureStorage {

    /**
     * Stores the API environment on which the SDK will operate.
     *
     * @param apiEnvironment The api environment to be stored.
     */
    fun setApiEnvironment(apiEnvironment: ApiEnvironment)

    /**
     * Retrieves the API environment
     */
    fun getApiEnvironment(): ApiEnvironment?

    /**
     * Stores the cloud authentication token.
     *
     * @param token The cloud authentication token to be stored.
     */
    fun addCloudAuthToken(token: String)

    /**
     * Retrieves the cloud authentication token.
     *
     * @return The stored cloud authentication token, or null if not found.
     */
    fun getCloudAuthToken(): String?

    /**
     * Stores the cloud refresh token.
     *
     * @param token The cloud refresh token to be stored.
     */
    fun addCloudRefreshToken(token: String)

    /**
     * Retrieves the cloud refresh token.
     *
     * @return The stored cloud refresh token, or null if not found.
     */
    fun getCloudRefreshToken(): String?

    /**
     * Stores the fusion host.
     *
     * @param host The fusion host to be stored.
     */
    fun setFusionHost(host: String)

    /**
     * Retrieves the fusion host.
     *
     * @return The stored fusion host, or null if not found.
     */
    fun getFusionHost(): String?

    /**
     * Stores the Fusion authentication token.
     *
     * @param token The Fusion authentication token to be stored.
     */
    fun addFusionAuthToken(token: String)

    /**
     * Retrieves the Fusion authentication token.
     *
     * @return The stored Fusion authentication token, or null if not found.
     */
    fun getFusionAuthToken(): String?

    /**
     * Stores the public key.
     *
     * @param publicKey The public key as a byte array.
     */
    fun addPublicKey(publicKey: ByteArray)

    /**
     * Retrieves the public key.
     *
     * @return The stored public key as a byte array, or null if not found.
     */
    fun getPublicKey(): ByteArray?

    /**
     * Stores the private key.
     *
     * @param privateKey The private key as a byte array.
     */
    fun addPrivateKey(privateKey: ByteArray)

    /**
     * Retrieves the private key.
     *
     * @return The stored private key as a byte array, or null if not found.
     */
    fun getPrivateKey(): ByteArray?

    /**
     * Stores the public key that has been verified via two-factor authentication.
     *
     * @param publicKey The key pair verification status
     */
    fun setKeyPairVerified(publicKey: ByteArray?)

    /**
     * Retrieves the verified public key and compares it against the stored public key.
     * Returns true if they match exactly, or false if either value is null or the keys don't match.
     *
     * @return The stored key pair verification status a boolean, or null if not found.
     */
    fun getKeyPairVerified(): ByteArray?

    /**
     * Stores the user ID.
     *
     * @param userId The user ID to be stored.
     */
    fun addUserId(userId: String)

    /**
     * Retrieves the user ID.
     *
     * @return The stored user ID, or null if not found.
     */
    fun getUserId(): String?

    /**
     * Stores the user email.
     *
     * @param email The user email to be stored.
     */
    fun addUserEmail(email: String)

    /**
     * Retrieves the user email.
     *
     * @return The stored user email, or null if not found.
     */
    fun getUserEmail(): String?

    /**
     * Stores the certificate chain.
     *
     * @param certificateChain The certificate chain as a list of strings.
     */
    fun addCertificateChain(certificateChain: List<String>)

    /**
     * Retrieves the certificate chain.
     *
     * @return The stored certificate chain as a list of strings, or null if not found.
     */
    fun getCertificateChain(): List<String>?

    /**
     * Clears all stored data.
     */
    fun clear()
}

internal expect fun createSecureStorage(applicationContext: ApplicationContext? = null): SecureStorage

File: doordeck-sdk/src/mingwMain/resources/csharp/Wrapper/SecureStorage.cs
using System.Runtime.InteropServices;
using System.Text;
using Doordeck.Headless.Sdk.Model;
using Doordeck.Headless.Sdk.Utilities;

namespace Doordeck.Headless.Sdk.Wrapper;

internal static class SecureStorage
{
    public static ISecureStorage? Implementation { get; set; }

    /// <summary>Entry names agreed with SecureStorageKeys on the Kotlin side.</summary>
    private const string ApiEnvironmentKey = "apiEnvironment";
    private const string CloudAuthTokenKey = "cloudAuthToken";
    private const string CloudRefreshTokenKey = "cloudRefreshToken";
    private const string FusionHostKey = "fusionHost";
    private const string FusionAuthTokenKey = "fusionAuthToken";
    private const string PublicKeyKey = "publicKey";
    private const string PrivateKeyKey = "privateKey";
    private const string KeyPairVerifiedKey = "keyPairVerified";
    private const string UserIdKey = "userId";
    private const string UserEmailKey = "userEmail";
    private const string CertificateChainKey = "certificateChain";

    public static class Delegates
    {
        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate void SetEntryDelegate(IntPtr key, IntPtr value);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate int GetEntryDelegate(IntPtr key, IntPtr buffer, int capacity);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        public delegate void ClearDelegate();
    }

    // Held in static fields for the lifetime of the process: the SDK keeps these pointers, so the
    // delegates they came from must stay reachable.
    private static readonly Delegates.SetEntryDelegate SetEntryDelegate = SetEntry;
    private static readonly Delegates.GetEntryDelegate GetEntryDelegate = GetEntry;
    private static readonly Delegates.ClearDelegate ClearDelegate = Clear;

    internal static void Register() =>
        Native.dd_set_secure_storage(
            Marshal.GetFunctionPointerForDelegate(SetEntryDelegate),
            Marshal.GetFunctionPointerForDelegate(GetEntryDelegate),
            Marshal.GetFunctionPointerForDelegate(ClearDelegate));

    private static void SetEntry(IntPtr key, IntPtr value)
    {
        if (Implementation is not {} implementation) return;

        var name = Marshal.PtrToStringUTF8(key);
        if (name == null) return;

        var entry = value == IntPtr.Zero ? null : Marshal.PtrToStringUTF8(value);

        switch (name)
        {
            case ApiEnvironmentKey when entry != null:
                implementation.SetApiEnvironment(Enum.Parse<ApiEnvironment>(entry));
                break;
            case CloudAuthTokenKey when entry != null:
                implementation.AddCloudAuthToken(entry);
                break;
            case CloudRefreshTokenKey when entry != null:
                implementation.AddCloudRefreshToken(entry);
                break;
            case FusionHostKey when entry != null:
                implementation.SetFusionHost(entry);
                break;
            case FusionAuthTokenKey when entry != null:
                implementation.AddFusionAuthToken(entry);
                break;
            case PublicKeyKey when entry != null:
                implementation.AddPublicKey(entry.DecodeBase64ToByteArray());
                break;
            case PrivateKeyKey when entry != null:
                implementation.AddPrivateKey(entry.DecodeBase64ToByteArray());
                break;
            // The only entry the SDK legitimately clears by storing a null value.
            case KeyPairVerifiedKey:
                implementation.SetKeyPairVerified(entry?.DecodeBase64ToByteArray());
                break;
            case UserIdKey when entry != null:
                implementation.AddUserId(new Guid(entry));
                break;
            case UserEmailKey when entry != null:
                implementation.AddUserEmail(entry);
                break;
            case CertificateChainKey when entry != null:
                implementation.AddCertificateChain(entry.StringToCertificateChain());
                break;
        }
    }

    private static int GetEntry(IntPtr key, IntPtr buffer, int capacity)
    {
        if (Implementation is not {} implementation) return -1;

        var name = Marshal.PtrToStringUTF8(key);
        if (name == null) return -1;

        var entry = name switch
        {
            ApiEnvironmentKey => implementation.GetApiEnvironment()?.ToString(),
            CloudAuthTokenKey => implementation.GetCloudAuthToken(),
            CloudRefreshTokenKey => implementation.GetCloudRefreshToken(),
            FusionHostKey => implementation.GetFusionHost(),
            FusionAuthTokenKey => implementation.GetFusionAuthToken(),
            PublicKeyKey => implementation.GetPublicKey()?.EncodeByteArrayToBase64(),
            PrivateKeyKey => implementation.GetPrivateKey()?.EncodeByteArrayToBase64(),
            KeyPairVerifiedKey => implementation.GetKeyPairVerified()?.EncodeByteArrayToBase64(),
            UserIdKey => implementation.GetUserId()?.ToString(),
            UserEmailKey => implementation.GetUserEmail(),
            CertificateChainKey => implementation.GetCertificateChain()?.CertificateChainToString(),
            _ => null
        };

        if (entry == null) return -1;

        // UTF-8 to match Kotlin/Native's toKString on the other side.
        var bytes = Encoding.UTF8.GetBytes(entry);

        // A null buffer, or one too small, is how the caller asks for the length before allocating.
        if (buffer != IntPtr.Zero && capacity > bytes.Length)
        {
            Marshal.Copy(bytes, 0, buffer, bytes.Length);
            Marshal.WriteByte(buffer, bytes.Length, 0);
        }

        return bytes.Length;
    }

    private static void Clear() => Implementation?.Clear();
}